beam-sqlite/Database/Beam/Sqlite/Connection.hs

```diff
@@ -335,11 +333,64 @@ instance MonadBeam Sqlite SqliteM where
                                 Nothing -> pure Nothing
                                 Just (BeamSqliteRow row) -> pure row
                runReaderT (runSqliteM (action nextRow')) (logger, conn)
-  runReturningMany SqliteCommandInsert {} _ =
-      fail . mconcat $
-      [ "runReturningMany{Sqlite}: sqlite does not support returning "
-      , "rows from an insert, use Database.Beam.Sqlite.insertReturning "
-      , "for emulation" ]
+  runReturningMany (SqliteCommandInsert (SqliteInsertSyntax tbl fields vs onConflict)) action
+    | SqliteInsertExpressions es <- vs, any (any (== SqliteExpressionDefault)) es =
+      -- SQLite's handling of default values differs from other DBMses because
+      -- it lacks support for DEFAULT. In order to insert a default value in a column,
+      -- the column's name should be omitted from the INSERT statement.
+      --
+      -- This is problematic if you insert multiple rows, some of which have defaults;
+      -- you must use multiple INSERT statements. This is what we do below.
+      --
+      -- However, to respect the 'runReturningMany' interface, be must accumulate the
+      -- results of all those inserts into an 'IORef [a]', and then feed the results
+      -- incrementally to 'action'.
+      SqliteM $ do
+        (logger, conn) <- ask
+        resultsRef <- liftIO (newIORef [])
+        forM_ es $ \row -> do
+          -- RETURNING is only supported by SQLite 3.35+, which requires direct-sqlite 2.3.27+
+          let returningClause = emit " RETURNING " <> commas (map quotedIdentifier fields)
+              (insertFields, insertRow) = unzip $ filter ((/= SqliteExpressionDefault) . snd) $ zip fields row
+              SqliteSyntax cmd vals = formatSqliteInsertOnConflict tbl insertFields (SqliteInsertExpressions [ insertRow ]) onConflict <> returningClause
+              cmdString = BL.unpack (toLazyByteString (withPlaceholders cmd))
```

Comment:
Only a minor comment (that probably should be punted to a separate ticket), but: would it be possible to instead classify the rows by which fields have DEFAULT values? For example, if all the rows to be inserted have DEFAULT in the same position, then we don't need to do this, right?

Reply:
Yes that's very true. 
I considered doing something like it, but I started with this simpler implementation instead because the order of insertion is preserved this way. If you grouped rows by the positions of their default values, then you would have to reconstruct the return order.
This isn't impossible, but this is significantly more complex. As you mention, this could be punted to a new ticket.

Actually, I see you have made a prototype implementation for this below; would you like to submit a PR?